tool/lib/license_utils.dart

```diff
@@ -4,46 +4,126 @@
 
 import 'package:path/path.dart' as p;
 import 'package:yaml/yaml.dart';
 
+/// This class contains config file related business logic for
+/// [license_utils.dart].
+///
+/// The [LicenseConfig] reads in config data from a YAML file into an
+/// object that can be used to update license headers.
+/// 
+/// Sample config file:
+/// ---
+/// # sequence of license text strings that should be matched against at the top of a file and removed. <value>, which normally represents a date, will be stored.
+/// remove_licenses:
+///   - |
+///     // This is some <value1> multiline license
+///     // text that should be removed from the file.
+///   - |
+///     /* This is other <value2> multiline license
+///     text that should be removed from the file. */
+///   - |
+///     # This is more <value3> multiline license
+///     # text that should be removed from the file.
+///   - |
+///     // This is some multiline license text to
+///     // remove that does not contain a stored value.
+/// # sequence of license text strings that should be added to the top of a file. {value} will be replaced.
+/// add_licenses: 
+///   - |
+///     // This is some <value1> multiline license
+///     // text that should be added to the file.
+///   - |
+///     # This is other <value3> multiline license
+///     # text that should be added to the file.
+///   - |
+///     // This is some multiline license text to
+///     // add that does not contain a stored value.
+/// # defines which files should have license text added or updated.
+/// update_paths:
+///   # path(s) to recursively check for files to remove/add license
+///   include:
+///       - /repo_root
+///   # path(s) to recursively check for files to ignore
+///   exclude:
+///     # exclude everything in the /repo_root/sub_dir1 directory
+///     - /repo_root/sub_dir1/
+///     # exclude the given files
+///     - /repo_root/sub_dir2/exclude1.ext1
+///     - /repo_root/sub_dir2/sub_dir3/exclude2.ext2
+///   file_types:
+///     # extension
+///     ext1:
+///       # one or more indices of remove_licenses to remove
+///       remove:
+///         - 0
+///         - 1
+///       # index of add_licenses to add
+///       add: 0
+///     ext2:
+///       remove:
+///         - 2
+///       add: 1
 class LicenseConfig {
+  /// Sequence of license text strings that should be matched against at the top
+  /// of a file and removed.
+  late final YamlList removeLicenses;
+
+  /// Sequence of license text strings that should be added to the top of a
+  /// file.
+  late final YamlList addLicenses;
+
+  /// Path(s) to recursively check for file to remove/add license text
+  late final YamlList includePaths;
+
+  /// Path(s) to recursively check for files to ignore
+  late final YamlList excludePaths;
+
+  /// Contains the extension (without a '.') and the associated indices
+  /// of [removeLicenses] to remove and index of [addLicenses] to add for the
+  /// file type.
+  late final YamlMap fileTypes;
```

Comment:
these no longer need to be late now that `LicenseConfig.fromYamlFile` is a factory constructor.

also, style-ordering-nit: order as follows:
// constructor
// factory constructors
// final fields